handler/src/main/java/io/netty/handler/ssl/SslHandler.java

```diff
@@ -589,14 +597,33 @@ private SSLEngineResult wrap(ByteBufAllocator alloc, SSLEngine engine, ByteBuf i
             throws SSLException {
         ByteBuf newDirectIn = null;
         try {
-            final ByteBuffer in0;
+            int readerIndex = in.readerIndex();
+            int readableBytes = in.readableBytes();
+
+            // We will call SslEngine.wrap(ByteBuffer[], ByteBuffer) to allow efficient handling of
+            // CompositeByteBuf without force an extra memory copy when CompositeByteBuffer.nioBuffer() is called.
+            final ByteBuffer[] in0;
             if (in.isDirect() || !wantsDirectBuffer) {
-                in0 = in.nioBuffer();
+                // As CompositeByteBuf.nioBufferCount() can be expensive (as it needs to check all composed ByteBuf
+                // to calculate the count) we will just assume a CompositeByteBuf contains more then 1 ByteBuf.
+                // The worst that can happen is that we allocate an extra ByteBuffer[] in CompositeByteBuf.nioBuffers()
+                // which is better then walking the composed ByteBuf in most cases.
+                if (!(in instanceof CompositeByteBuf) && in.nioBufferCount() == 1) {
```

Comment:
@trustin - I created https://github.com/netty/netty/issues/3306 to address this issue.  There is more explanation/justification in the issue description.